### PR TITLE
Display popular modules on averaged downloads per day

### PR DESCRIPTION
**Pull request by Firesphere**
By default, on page load, show the average downloads per day since it first showed up in Packagist.

The user has the option to switch to total amount of downloads, which will do an ajax call to retrieve the rendered HTML for the total amount.
Or switch back, of course.

Sorting is done on the full float number, on the frontend, the number is formatted to 2 behind the separator.

Reasoning for using ajax for changing the content is obvious to me. It would be way to obtrusive to fully reload the page.

Note, it is _not_ applied to the add-ons page, which still only sorts by total downloads. I've been looking at that, but it needs more work to get it to work properly.

### CODE REVIEW

**Comment by Firesphere** [comment]
sorting-defaults-update

How it works. Sorry, my gif starts at the total downloads, it doesn't on load, but I'd been clicking around a bit before starting recording.

File: mysite/code/controllers/HomeController.php [hunk 1]
```diff
@@ -28,13 +28,40 @@ public function Link() {
 		return Director::baseURL();
 	}
 
-	public function PopularAddons($limit = 10) {
+    /**
+     * @param int $limit
+     * @return DataList|SS_Limitable default to Relative addons
+     */
+	public function PopularAddons($limit = 10)
```
Comment (robbieaverill) [inline mysite/code/controllers/HomeController.php]: Indentation is a bit whack

File: mysite/code/controllers/HomeController.php [hunk 2]
```diff
@@ -28,13 +28,40 @@ public function Link() {
...
+     */
+	public static function AbsolutePopularAddons($limit = 10) {
 		return Addon::get()
 			->sort('Downloads', 'DESC')
-			->exclude('Name', $this->config()->popular_blacklist)
+			->exclude('Name', self::config()->popular_blacklist)
 			->limit($limit);
 	}
 
+	public static function RelativePopularAddons($limit = 10)
```
Comment (robbieaverill) [inline mysite/code/controllers/HomeController.php]: Here too

File: mysite/code/dataobjects/Addon.php [hunk 3]
```diff
@@ -163,12 +163,30 @@ public function getDateCreated() {
...
+	 * @return ArrayData
 	 */
 	public function HelpfulRobotData()
 	{
 		$data = json_decode($this->HelpfulRobotData, true);
 
 		return new ArrayData($data["inspections"][0]);
 	}
+
+	public function addonAge()
```
Comment (robbieaverill) [inline mysite/code/dataobjects/Addon.php]: And here

File: mysite/code/dataobjects/Addon.php [hunk 4]
```diff
@@ -163,12 +163,30 @@ public function getDateCreated() {
...
+        $released = new DateTime($this->Released);
+        return $date->diff($released);
+    }
+
+    public function getRelativePopularity()
+    {
+        return (int)$this->Downloads / (int)$this->addonAge()->days;
+    }
+
+    public function relativePopularityFormatted()
```
Comment (robbieaverill) [inline mysite/code/dataobjects/Addon.php]: Could you add doc blocks to these methods?

**Comment by robbieaverill** [comment]
Thanks - as mentioned in Slack, for me (OSX Sierra + Chrome 55.0.2883.95) the XHR response is identical when you choose either option from the dropdown, so the stats don't change.

The HomeController results seem to be fine in that "per day" has a `Score` which is "x per day", and overall doesn't contain the `Score` property. Perhaps the template is being cached?

**Comment by robbieaverill** [comment]
@Firesphere thanks for the update. LGTM.

**Comment by sminnee** [comment]
I saw this finally went live, which is awesome!

A bit of feedback:

 * We probably don't need the figure to 2-decimal places
 * Would it be better to use "downloads in the last 30 days" as the basis for this instead of downloads all time? This would give its current, rather than average, popularity.
 * It would be much better if the sorting could be done by a database column. You could calculate the figure and write it to the database as part of the import?

**Comment by Firesphere** [comment]
To answer a bit to your feedback @sminnee:
- The 2 decimals is arbitrarily chosen. I don't think no decimals at all is a good idea, but that's a matter of preference I guess
- Originally, based on @dhensby his feedback, I went with using the total downloads, not the last thirty days.
- Sounds like a good idea as long as it doesn't slow down the updating process too much?

**Comment by sminnee** [comment]
The use of total downloads / total days will keep something that used to be popular but is no longer in the list, but it will penalise a module that was unknown for a long time and has recently become more popular. 30 days will do the opposite.

It depends what we're aiming for I guess; not sure what others think?

**Comment by Firesphere** [comment]
I agree, I'd like @dhensby his thought on this?

**Comment by SpiritLevel** [comment]
You could get fancy and use an exponentially weighted moving average which puts a higher weight on more recent downloads and a smaller weight on downloads further in the past:

`M(t) = [ a M(t-1) + a^2 M(t-2) + ... + a^nM(t-n) ] * (1-a)/(a-a^(n+1))`

where `M(t)` is the "average" number of downloads in month `t` and where `0<a<1` controls the weighting of the past relative to the present. Small `a` emphasises the near past while `a` near 1 makes the past as important as the present. I would suggest a 24 month window and `a=0.8` in which case all monthly download amounts older than two years are ignored in calculating the average and the weights for the most recent 24 months are 0.2, 0.16,  0.13, 0.082, 0.065, 0.054, 0.048, ... , 0.0001.

**Comment by dhensby** [comment]
My thoughts are that both using the last 30 days and all time are overly simplified but that using the last 30 days had the potential to hide those that have stood the test of time. This was all based on the assumption we didn't have the time/desire to implement a more "intelligent" solution such as the one suggested by @SpiritLevel

**Pull request merged**